### PR TITLE
Mouse input and 60fps fixes

### PR DESCRIPTION
**Pull request by mercury501**
I'm here if you have suggestions/critiques!

### CODE REVIEW

**Comment by mercury501** [comment]
BTW I'll be adding fixes requested by @Polymega [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/28#issuecomment-1325738277) in the coming days, so no rush in merging it! I'll update when it's done

**Comment by mercury501** [comment]
Ok I think we're done with cutscene fixes!

**Comment by Polymega** [comment]
Yep! I think everything is good to go for the PR now! 

A million thanks to you, Mercury, for taking the lead and creating the epic mouse input work, along with managing all the 60 fps work that's been going on! And a million more thanks to Murugo and Aero_ once again, as their efforts were also pivotal in getting us where we're at with this epic PR.

You all started 2023 off with a bang! Thank you so much. ❤️

File: Wrappers/dinput8/IDirectInputDevice8A.cpp [hunk 1]
```diff
@@ -92,28 +92,50 @@ HRESULT m_IDirectInputDevice8A::Acquire()
 {
 	Logging::LogDebug() << __FUNCTION__ << " (" << this << ")";
 
+	LPDIDEVICEINSTANCEA deviceInfo = new DIDEVICEINSTANCEA;
```
Comment (elishacloud) [inline Wrappers/dinput8/IDirectInputDevice8A.cpp:95]: There is a memory leak here.  You create a new device without ever deleting it later.  Instead this line should be something like this:

```c++
DIDEVICEINSTANCEA deviceInfo;
```

File: sh2-enhce.vcxproj.bak [hunk 2]
```diff
@@ -0,0 +1,339 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (elishacloud) [inline sh2-enhce.vcxproj.bak:1]: This file is not needed.  You should delete this file.

File: sh2-enhce.vcxproj.filters.bak [hunk 3]
```diff
@@ -0,0 +1,734 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (elishacloud) [inline sh2-enhce.vcxproj.filters.bak:1]: This file is not needed.  You should delete this file.

**Review by elishacloud — CHANGES_REQUESTED** [review]
I finished my first review and requested some changes that need to be made.

**Comment by elishacloud** [comment]
One more thing I noticed is that when running the DC version of the game I am getting errors on the log file.  I guess this issue is an incorrect address on line 1732 of `Common.cpp`.  See below:

```txt
25524 14:44:43.072 Patching water animation speed...
25524 14:44:43.072 Patching flashlight speed...
25524 14:44:43.072 CheckMemoryAddress Error: memory address not found!
25524 14:44:43.073 SearchAndGetAddresses searching for memory address! Found = 0050A1EA
25524 14:44:43.073 Patching low health indicator flashing speed...
25524 14:44:43.073 Patching flame staircase lighting...
```

In this code here:

```c++
int16_t *GetFlashlightOnSpeedPointer()
{
	if (FlashlightOnSpeedAddr)
	{
		return FlashlightOnSpeedAddr;
	}
	
	// Get Items Collected address
	constexpr BYTE FlashlightOnSpeedSearchBytes[]{ 0x74, 0x11, 0x6A, 0x00, 0xE8, 0xAD, 0x79 };
	int16_t *FlashlightOnSpeed = (int16_t*)SearchAndGetAddresses(0x0050A59A, 0x0050A8CA, 0x0050A8CA, FlashlightOnSpeedSearchBytes, sizeof(FlashlightOnSpeedSearchBytes), 0x23);

	// Checking address pointer
	if (!FlashlightOnSpeed)
	{
		Logging::Log() << __FUNCTION__ << " Error: failed to find Flashlight On Speed address!";
		return nullptr;
	}

	FlashlightOnSpeedAddr = (int16_t*)((DWORD)FlashlightOnSpeed);

	return FlashlightOnSpeedAddr;
}
```

File: Patches/Common.cpp [hunk 4]
```diff
@@ -722,6 +766,40 @@ BYTE *GetPauseMenuButtonIndexPointer()
...
+BYTE GetPauseMenuQuitIndex()
+{
+	BYTE *PauseMenuIndex = GetPauseMenuQuitIndexPointer();
+
+	return (PauseMenuIndex) ? *PauseMenuIndex : 0;
+}
+
+BYTE *GetPauseMenuQuitIndexPointer()
+{
+	return (BYTE*)0x009326B0;
```
[the remark this answers is not in the document]
Comment (mercury501) [reply Patches/Common.cpp:778]: Forgot to dial in the address, fixed in next commit

File: Common/Settings.h [hunk 5]
```diff
@@ -26,14 +26,17 @@
 	visit(DPadMovementFix, true) \
 	visit(DynamicResolution, true) \
 	visit(EnableDebugOverlay, true) \
+	visit(EnableEnhancedMouse, true) \
 	visit(EnableInfoOverlay, true) \
 	visit(EnableMenuTest, false) \
 	visit(EnableMenuTestIGT, true) \
+	visit(EnableMouseWheelSwap, false) \
```
[the remark this answers is not in the document]
Comment (mercury501) [reply Common/Settings.h:33]: This is correct, it should be disabled by default, right @Polymega ?

File: Launcher/config.xml [hunk 6]
```diff
@@ -463,6 +485,28 @@
...
+          <Value name="Disable">0</Value>
+          <Value name="Enable" default="true">1</Value>
+        </Choices>
+      </Feature>
+
+    <Feature name="EnableMouseWheelSwap">
+      <Title>Enable mouse weapon swap</Title>
+      <Description>Enables Mousewheel to swap weapons</Description>
+      <Choices type="check">
+        <Value name="Disable" default="false">0</Value>
```
[the remark this answers is not in the document]
Comment (mercury501) [reply Launcher/config.xml:503]: This was changed but got reverted while merging, fixed it

File: Launcher/config.xml [hunk 7]
```diff
@@ -746,6 +790,24 @@
         </Choices>
       </Feature>
 
+      <Feature name="MemoScreenFix">
+        <Title>Fixe navigation for the memo screen</Title>
```
[the remark this answers is not in the document]
Comment (mercury501) [reply Launcher/config.xml:794]: Done!

File: Patches/Common.cpp [hunk 8]
```diff
@@ -58,9 +59,52 @@ int16_t *ItemsCollectedAddr;
...
+float* FrametimeAddr;
+DWORD* MeatLockerFogFixOneAddr;
+DWORD* MeatLockerFogFixTwoAddr;
+DWORD* MeatLockerHangerFixOneAddr;
+DWORD* MeatLockerHangerFixTwoAddr;
+BYTE* ClearTextAddr;
+float* MeetingMariaCutsceneFogCounterOneAddr;
+float* MeetingMariaCutsceneFogCounterTwoAddr;
+float* RPTClosetCutsceneMannequinDespawnAddr;
+float* RPTClosetCutsceneBlurredBarsDespawnAddr;
```
[the remark this answers is not in the document]
Comment (mercury501) [reply Patches/Common.cpp:100]: I stopped adding the initialization for some reason, fixed all of them!

File: Wrappers/dinput8/IDirectInputDevice8A.cpp (hunk 1 above)
Comment (mercury501) [reply Wrappers/dinput8/IDirectInputDevice8A.cpp:95]: Absolutely right, fixed it!

**Comment by mercury501** [comment]
> One more thing I noticed is that when running the DC version of the game I am getting errors on the log file. I guess this issue is an incorrect address on line 1732 of `Common.cpp`. See below:
> 
> ```
> 25524 14:44:43.072 Patching water animation speed...
> 25524 14:44:43.072 Patching flashlight speed...
> 25524 14:44:43.072 CheckMemoryAddress Error: memory address not found!
> 25524 14:44:43.073 SearchAndGetAddresses searching for memory address! Found = 0050A1EA
> 25524 14:44:43.073 Patching low health indicator flashing speed...
> 25524 14:44:43.073 Patching flame staircase lighting...
> ```
> 
> In this code here:
> 
> ```c++
> int16_t *GetFlashlightOnSpeedPointer()
> {
> 	if (FlashlightOnSpeedAddr)
> 	{
> 		return FlashlightOnSpeedAddr;
> 	}
> 	
> 	// Get Items Collected address
> 	constexpr BYTE FlashlightOnSpeedSearchBytes[]{ 0x74, 0x11, 0x6A, 0x00, 0xE8, 0xAD, 0x79 };
> 	int16_t *FlashlightOnSpeed = (int16_t*)SearchAndGetAddresses(0x0050A59A, 0x0050A8CA, 0x0050A8CA, FlashlightOnSpeedSearchBytes, sizeof(FlashlightOnSpeedSearchBytes), 0x23);
> 
> 	// Checking address pointer
> 	if (!FlashlightOnSpeed)
> 	{
> 		Logging::Log() << __FUNCTION__ << " Error: failed to find Flashlight On Speed address!";
> 		return nullptr;
> 	}
> 
> 	FlashlightOnSpeedAddr = (int16_t*)((DWORD)FlashlightOnSpeed);
> 
> 	return FlashlightOnSpeedAddr;
> }
> ```

I must've copy pasted the DC address wrong, fixed it!

**Comment by mercury501** [comment]
I'll go through the code again tomorrow to check my last commit, thanks for your inputs @elishacloud

File: Common/Settings.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Polymega) [reply Common/Settings.h:33]: Let's enable it, please! I was originally worried people would spam the mouse wheel, make James look goofy, then complain about that. But it's up to the person to "Mouse Wheel Responsibly."

[the remark this answers is not in the document]
Comment (mercury501) [reply Common/Settings.h]: done!

**Review by elishacloud — APPROVED** [review]

**Comment by Polymega** [comment]
@IlDucci 
My memory is a bit hazy on this, as a lots been going on for a little while now. But when you have the opportunity, may I ask you to look over the Spanish XML file for the config tool to see if it needs any new features to be translated? I also just went in to clean up or clarify some features. May I ask you to look over these changes for the Spanish version?

https://github.com/mercury501/Silent-Hill-2-Enhancements/commit/057208b8a0b1b4e035e7105278e9f40bf6627977

Thank you!

**Comment by elishacloud** [comment]
Currently none of these new features have been added to the Spanish XML file yet.  All of them need to be added.

**Comment by elishacloud** [comment]
I just checked in the last change.  Everything looks good at this point.  Once the Spanish XML is done this will be ready to merge.

**Comment by Polymega** [comment]
Very exciting! Upon reviewing the code, you probably picked up on this, so for anyone else lurking this thread:

Mercury's PR is doing a lot more than including mouse input and 60 fps features. This huge PR does all of the following:

- Mouse input support (character movement, menu navigation improvements, and more)
- 60 fps mode with high framerate fixes
- Independent FMV framerate feature
- A fix for assigning functions to non-numpad 0-9 (which are typically hardcoded for weapons only)
- A fix for `FixInventoryBGM`
- Improvements for `RemoveEnvironmentFlicker`
- Improvements for `GameLoadFix`
- Improvements for `ClosetCutsceneFix`
- All new `QuickSaveTweaks` feature
- Possibly other things I'm forgetting

... this is a big one!

**Comment by elishacloud** [comment]
Thanks for all the work @mercury501!

Here is the latest build with all the merged updates:  [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10375499/d3d8.zip)

**Comment by MasterDarkseid** [comment]
> Thanks for all the work @mercury501!
> 
> Here is the latest build with all the merged updates: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10375499/d3d8.zip)

can he be freely be tested.

**Comment by mercury501** [comment]
@MasterDarkseid let's see what @Polymega  as to say, in any for testing I think it would be better to use a build with both 60 fps fixes and mouse input, to catch two birds with one stone

**Comment by MasterDarkseid** [comment]
@mercury501 isn't both merged now ?

**Comment by mercury501** [comment]
Yeah my bad, thought this was the 60 fps issue, not the actual pr

**Comment by mercury501** [comment]
Anyways I think its good to test, the major implementation is there, there's just a small bug fixed in the open pr

**Comment by MasterDarkseid** [comment]
> Anyways I think its good to test, the major implementation is there, there's just a small bug fixed in the open pr

Before doing so is there any warnings or any change to do or everything is set in D3D8

**Comment by mercury501** [comment]
Iirc all of the new features should be enabled by default, can't check since I'm not at the pc, but if you find quicksavetweaks = 1 in the ini, on Ctrl + g you see 60 fps and mouse turns James, you should be good

**Comment by MasterDarkseid** [comment]
> Iirc all of the new features should be enabled by default, can't check since I'm not at the pc, but if you find quicksavetweaks = 1 in the ini, on Ctrl + g you see 60 fps and mouse turns James, you should be good

Just got a softlock at the clock in the appartment.

I've examined the clock time then used the key and the "I used the clock key" remain stuck only the shift button seems to work as he put james in guard pose but don't do anything else.

**Comment by Polymega** [comment]
Can you consistently replicate this issue?

**Comment by MasterDarkseid** [comment]
> Can you consistently replicate this issue?

I will be back i gonna do a series of tests, i'll report back in 15m max.

**Comment by MasterDarkseid** [comment]
Never had issues before that D3D8. So i guess 60 FPS might be the cause, more testing needed.

**Comment by MasterDarkseid** [comment]
Another edit: in fact could it be the fade out changes you did (If it's already applied) Otherwise ignore the post he might be another issue.

**Comment by Polymega** [comment]
Can you record video of yourself provoking this bug?

**Comment by MasterDarkseid** [comment]
> Can you record video of yourself provoking this bug?

Here replicated it a third time.

https://twitter.com/PrimeDarkseid/status/1614781976539537409

There you go easy reproducing.

**Comment by Polymega** [comment]
This looks to be only happening _after_ an in-game text selection ("YES/NO"). If you set `QuickSaveTweaks = 0` in the INI, does the bug still happen?

**Comment by MasterDarkseid** [comment]
> This looks to be only happening _after_ an in-game text selection ("YES/NO"). If you set `QuickSaveTweaks = 0` in the INI, does the bug still happen?
> 
> This bug looks to be provoked by pressing Esc instantly or near instantly at the same time you press "Accept" on a YES/NO selection.

Not really, just text prompt skip can also cause this. Also the D3D8 seems to have been shipped with no QuickSaveTweaks commands at all. 60 FPS, Mouse, FMV fix are all there but no sign QuickSaveTweaks in the files.

So i take the thing is set on 0 as he don't exist in the D3D8

**Comment by MasterDarkseid** [comment]
Nevermind he does appear in the D3D8 text but not in the config.

**Comment by MasterDarkseid** [comment]
And yes he still happen even with the QuickSaveTweaks set on 0.

**Comment by Polymega** [comment]
I've figured out how to consistently replicate the bug. It is from `QuickSaveTweaks` and to provoke the bug is simple:

Anywhere there is an in-game text selection (example: "YES/NO"), press the Esc key while this selection is active to softlock the game:

https://user-images.githubusercontent.com/27395260/212577467-3512ff6f-d5d8-4ac7-8e8b-48de79b1ce00.mp4

We'll look into this. Thanks for catching that!

**Comment by mercury501** [comment]
Implemented @Polymega's suggestion, tested on the catacombs drop down prompt and didn't softlock, here's the build:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10425315/d3d8.zip)
Great job @MasterDarkseid

**Comment by MasterDarkseid** [comment]
> Implemented @Polymega's suggestion, tested on the catacombs drop down prompt and didn't softlock, here's the build: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10425315/d3d8.zip) Great job @MasterDarkseid

I test it right away.

**Comment by MasterDarkseid** [comment]
@Polymega @mercury501 I'm halfway done and did try to break it a few timesand he didn't break.

I'll update as i get though the game, while observing all the 60 Fixes. I noticed that the game harder due to fixed hitboxes. Always i did noticed the footstep being normal speed now while traveling to town.

**Comment by MarioTainaka** [comment]
> > Implemented @Polymega's suggestion, tested on the catacombs drop down prompt and didn't softlock, here's the build: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10425315/d3d8.zip) Great job @MasterDarkseid
> 
> I test it right away.

@Polymega  would it be okay if I did a casual playthrough to test this out?

**Comment by MarioTainaka** [comment]
> Of course! You never have to ask. The more who test, the merrier. Thanks

Alright grabbed the latest ini and dll and on my way.

**Comment by MarioTainaka** [comment]
> Of course! You never have to ask. The more who test, the merrier. Thanks

Alright so first impressions: You're all wizards. You really did fix most of the issues! @MasterDarkseid  seems to have caught most of them. I've noticed two issues so far the first one is that sometimes going through text when pressing the A Button seems sensitive sometimes. Like you'll skip multiple lines if you press A once or tap it slightly 

Also the second one seems to be that the lying figure seems to do a bit more damage when it's crawling compared to when you're playing 30 (might be imagining this one tho)

**Comment by MarioTainaka** [comment]
Alright first playthrough done. Had to stop and get back to work but yeah besides those 3 issues I mentioned this is a smooth experience. I don't think I can go back to 30.

**Comment by Polymega** [comment]
Glad you enjoyed the playthrough!

> Also the second one seems to be that the lying figure seems to do a bit more damage when it's crawling compared to when you're playing 30 (might be imagining this one tho)

This is a valid thing to check, as I think the damage is applied to James during the frames that the Lying Figure is in contact with him. So at twice the framerate this _could_ be a concern. 

It's hard to do a true side-by-side test for this, due to the creatures' random scutter patterns, but it looks like their might be some minor discrepancies:

https://user-images.githubusercontent.com/27395260/212793608-00f933ee-f5e3-4d1f-97fc-4d3d2d51ce63.mp4

^ Tested on Hard Action. 30 FPS seems to decrease health by 5/10 point increments, and 60 FPS can do 5/7.5/10 point increments. Depending on the frame that the Lying Figure is no longer "touching" James, the total damage accrued will either be the same or off by 2.5/5/10 points.

Again though, the tricky thing here is that it's kind of hard to reliably/consistently test this due to the chaotic nature of how they behave on the ground. There's no saying whether or not the Lying Figure was touching James for the same amount of time for each of those tests. @AeroWidescreen @Murugo thoughts?

**Comment by MarioTainaka** [comment]
> Glad you enjoyed the playthrough!
> 
> > Also the second one seems to be that the lying figure seems to do a bit more damage when it's crawling compared to when you're playing 30 (might be imagining this one tho)
> 
> This is a valid thing to check, as I think the damage is applied to James during the frames that the Lying Figure is in contact with him. So at twice the framerate this _could_ be a concern.
> 
> It's hard to do a true side-by-side test for this, due to the creatures' random scutter patterns, but it looks like their might be some minor discrepancies:
> 
>  NEW.mp4 
> ^ Tested on Hard Action. 30 FPS seems to decrease health by 5/10 point increments, and 60 FPS can do 5/7.5/10 point increments. Depending on the frame that the Lying Figure is no longer "touching" James, the total damage accrued will either be the same or off by 2.5/5/10 points.
> 
> Again though, the tricky thing here is that it's kind of hard to reliably/consistently test this due to the chaotic nature of how they behave on the ground. There's no saying whether or not the Lying Figure was touching James for the same amount of time for each of those tests. @AeroWidescreen @Murugo thoughts?

one test I kinda set up was in the blue creek apartments at full health and stamina it's very easy there to get em just crawl towards James in a straight pattern.

Oh one more thing. Would you mind checking the values for the final boss's strangle attack? I feel like the damage is alot higher then it should be.

**Comment by Polymega** [comment]
It's hard to test this, but there does seem to be _something_ in place for the scuttle damage. Here it is without any 60 FPS fixes:

https://user-images.githubusercontent.com/27395260/212799382-266a4743-7078-4c00-8329-adcaafee9cd9.mp4

[I monitored Murugo's `005359E8` address](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/28#issuecomment-1306880038) but it wasn't being called during these attacks, as far as I could tell. So I'm not sure what fix is in effect to help with this?

**Comment by MarioTainaka** [comment]
> It's hard to test this, but there does seem to be _something_ in place for the scuttle damage. Here it is without any 60 FPS fixes:
> 
>  NEW3.mp4 
> [I monitored Murugo's `005359E8` address](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/28#issuecomment-1306880038) but it wasn't being called during these attacks, as far as I could tell. So I'm not sure what fix is in effect to help with this?
> 
> > Oh one more thing. Would you mind checking the values for the final boss's strangle attack? I feel like the damage is alot higher then it should be.
> 
> This is fine. It's the same damage over time at both framerates. Same goes for the hold attacks by the Flesh Lips, Abstract Daddy, and the first Pyramid Head boss (when he choke holds you).

Kay, I'll be doing another playthrough tonight.

**Comment by MasterDarkseid** [comment]
Yeah, i'm 100% positive that the game got harder actually. Not that i complain i like hard/skilled stuff. But, as it's probably not intended, i suppose some fixes might be needed. He seems like that the Lying figure in particular seems to be affected.

**Comment by Polymega** [comment]
> Yeah, i'm 100% positive that the game got harder actually.

Enemies do the same amount of damage no matter the framerate. And, upon our fix for it, "stun lock" damage is now also the same (I tested this with all enemies that can stun lock you). With the game "feeling" harder, it's likely either a placebo effect or the following:

- Do enemies attempt attacks more often at a higher framerate?
   - I don't think they do. If they do, I'm not sure if that can be addressed. Someone could test this, although it might be tricky as there's probably an RNG roll for when enemies attack next.
- Does the scuttle of downed Lying Figures hurt you more at a higher framerate?
   - I tried testing this the other day (see a few posts up). This one is hard to test due to their chaotic nature once on the ground. I don't _think_ they hurt more when our framerate fixes are active, but I'm not sure.

**Comment by Polymega** [comment]
Hmm, I'm not sure what you mean by stacks? Damage-over-time is now the same for enemies that grapple you, and damage per singular attack (for non-grapple attacks) doesn't need adjustments as it was never frame dependent.

With SH2, the enemies don't deal more damage on different Action difficulties; they just attack more often at higher Action difficulties.

**Comment by MarioTainaka** [comment]
> Hmm, I'm not sure what you mean by stacks? Damage-over-time is now the same for enemies that grapple you, and damage per singular attack (for non-grapple attacks) doesn't need adjustments as it was never frame dependent.
> 
> With SH2, enemies don't deal different damage amounts on different Action difficulties; they just attack more often at higher Action difficulties.

Maybe it's just the 60FPS playing tricks on me.

super off topic but there's a bug with the latest public build available on the Enhanced Edition website:

the ADX\AFS files aren't being redirected. Sddata.bin is though. With the latest 60FPS fixes build everything is at it should be.

**Comment by MarioTainaka** [comment]
@Polymega  So here's some proof. I've put the latest public dll  through Reloaded II's file monitor VS the current dev dll. And sure enough the ADX and AFS files aren't being redirected but Sddata.bin is
[Latest dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442083/Latest.dll.logs.txt)
[Public Website dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442084/Public.Website.dll.logs.txt)

These logs show what gets loaded during the first two areas of the game.

**Comment by Murugo** [comment]
> [I monitored Murugo's `005359E8` address](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/28#issuecomment-1306880038) but it wasn't being called during these attacks, as far as I could tell. So I'm not sure what fix is in effect to help with this?

I tried testing this in Normal difficulty: the first hit always inflicts 5.0 damage regardless of framerate (instruction at `00535987`). For every subsequent frame, I do see `005359E8` being called instead which uses damage scaled by frame limit: 5.0 per frame in 30 FPS, and 2.5 per frame in 60 FPS (this is the "hold" damage fix).

I think it is possible for James to either take the same amount of damage or slightly more in 60 FPS depending on when he leaves the hitbox. Illustrated below: assume that the attack starts at `t = 0`. The tick rate in 60 FPS is `t = 0, 1, 2, 3 ...`, and in 30 FPS the tick rate is `t = 0, 2, 4, 8 ...`. If James leaves the hitbox on an "odd" frame (`t > 0, t % 2 == 1`), the total amount of damage inflicted will be +2.5 higher in 60 FPS. In practice though, the total damage will likely vary due to the random nature of the attack.

LyingFigureDamage

**Comment by Polymega** [comment]
Fantastic! Thanks for taking some time to confirm this, Murugo. (The visual aid is a nice touch, too!)

> For every subsequent frame, I do see 005359E8 being called instead which uses damage scaled by frame limit: 5.0 per frame in 30 FPS, and 2.5 per frame in 60 FPS (this is the "hold" damage fix).

Ah, so this is indeed being called and I missed it. Great then! Yeah, when I tested their scutter without any of our high framerate fixes I could definitely tell things were different. Good to know.

So, with Murugo's confirmation, we can say this is good/fixed as well!

**Comment by mercury501** [comment]
> @Polymega So here's some proof. I've put the latest public dll through Reloaded II's file monitor VS the current dev dll. And sure enough the ADX and AFS files aren't being redirected but Sddata.bin is [Latest dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442083/Latest.dll.logs.txt) [Public Website dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442084/Public.Website.dll.logs.txt)
> 
> These logs show what gets loaded during the first two areas of the game.

We are aware of this, infact there is a fix in this very PR, have you tried the latest build from this thread?

**Comment by mercury501** [comment]
Here is a build with all the latest fixes:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10448142/d3d8.zip)

**Comment by MarioTainaka** [comment]
> > @Polymega So here's some proof. I've put the latest public dll through Reloaded II's file monitor VS the current dev dll. And sure enough the ADX and AFS files aren't being redirected but Sddata.bin is [Latest dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442083/Latest.dll.logs.txt) [Public Website dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442084/Public.Website.dll.logs.txt)
> > These logs show what gets loaded during the first two areas of the game.
> 
> We are aware of this, infact there is a fix already in place, have you tried the latest build from this thread?

Yup! I'm actually using the latest build for 60FPS testing and that works fine as I said. I just wanted to check out the public branch just to test something and stumbled upon this issue by pure accident.

**Comment by mercury501** [comment]
> > > @Polymega So here's some proof. I've put the latest public dll through Reloaded II's file monitor VS the current dev dll. And sure enough the ADX and AFS files aren't being redirected but Sddata.bin is [Latest dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442083/Latest.dll.logs.txt) [Public Website dll logs.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10442084/Public.Website.dll.logs.txt)
> > > These logs show what gets loaded during the first two areas of the game.
> > 
> > 
> > We are aware of this, infact there is a fix already in place, have you tried the latest build from this thread?
> 
> Yup! I'm actually using the latest build for 60FPS testing and that works fine as I said. I just wanted to check out the public branch just to test something and stumbled upon this issue by pure accident.

All good then! Thank you for testing

**Comment by mercury501** [comment]
By the way, for this testing round we could move to

**Pull request merged**